cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,197 @@
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+      
+      // update validity
+      if(has_validity){         
+         // the final validity mask for this warp
+         int w_mask = __ballot_sync(w_active_mask, valid);
+         // only one guy in the warp needs to update the mask and count
+         if(w_lane_id == 0){
+            out.set_mask_word(w_cur, w_mask);            
+            w_valid_count += __popc(w_mask);
+         }
+      }      
+
+      // next grid
+      w_cur += blockDim.x * gridDim.x;
+      if(has_validity){
+         w_active_mask = __ballot_sync(w_active_mask, w_cur < w_end);
+      }
+   }
+
+   if(has_validity){
+      // sum all null counts across all warps
+      size_type b_valid_count = single_lane_block_sum_reduce<block_size, w_leader_lane>(w_valid_count);
+      // b_null_count will only be valid on thread 0
+      if(threadIdx.x == 0){
+         // using an atomic here because there are multiple blocks doing this work
+         atomicAdd(valid_count, b_valid_count);
+      }      
+   }
+}
+
+
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+*        using a filter function to select from lhs/rhs columns.
+*/
+/* ----------------------------------------------------------------------------*/
+struct copy_if_else_functor {
+   template <typename T, typename Filter>
+   void operator()(  column_view const& lhs,
+                     column_view const& rhs,
+                     Filter filter,
+                     mutable_column_view& out,
+                     rmm::mr::device_memory_resource *mr,                  
+                     cudaStream_t stream)
+   {
+      cudf::size_type num_els = warp_size * cudf::util::div_rounding_up_safe(lhs.size(), warp_size);
+      constexpr int block_size = 256;
+      cudf::experimental::detail::grid_1d grid{num_els, block_size, 1};
+
+      // device views
+      auto lhs_dv = column_device_view::create(lhs);
+      auto rhs_dv = column_device_view::create(rhs);
+      auto out_dv = mutable_column_device_view::create(out);
+      
+      if(lhs.nullable()){                           
+         rmm::device_scalar<cudf::size_type> valid_count{0, stream, mr};
+         
+         // call the kernel
+         copy_if_else_kernel<block_size, T, Filter, true><<<grid.num_blocks, block_size, 0, stream>>>(
+            *lhs_dv, *rhs_dv, filter, *out_dv, valid_count.data());
+
+         out.set_null_count(lhs.size() - valid_count.value());
```

Comment:
This doesn't actually modify the null count of the column you are viewing. It updates the null count of the _view_, not the `column`. So instead this functor should allocate the column so it's null count can be updated directly.

Reply:
Oh, wow good catch.  So is this just working because null_count() is lazily doing a recompute presumably during the equality checkin?

Reply:
>  So is this just working because null_count() is lazily doing a recompute presumably during the equality check in the test code?

Yep, pretty much. Any time you create a `mutable_view` from a `column`, it invalidates the internal `null_count` (for exactly this reason). So you'll always get correct results, it's just more performance optimal to update the null count in the column directly. 

I'd like to make it such that you _can_ modify the column's null count via a `mutable_view`, but it's a little tricky as it would require a `mutable_view` to accept a pointer (or reference) to it's null count, but I haven't figured out all the details for now.